controller/volume_controller.go

```diff
@@ -3734,50 +3734,49 @@ func (c *VolumeController) processMigration(v *longhorn.Volume, es map[string]*l
 		err = errors.Wrapf(err, "failed to process migration for %v", v.Name)
 	}()
 
-	if !isMigratableVolume(v) {
+	if !util.IsMigratableVolume(v) {
 		return nil
 	}
 
 	log := getLoggerForVolume(c.logger, v).WithField("migrationNodeID", v.Spec.MigrationNodeID)
 
-	// only process if volume is attached and running
-	if v.Spec.NodeID == "" || v.Status.CurrentNodeID == "" || len(es) == 0 {
-		return nil
-	}
-	if v.Status.Robustness != longhorn.VolumeRobustnessDegraded && v.Status.Robustness != longhorn.VolumeRobustnessHealthy {
-		log.Warnf("Skip the migration processing since the volume current robustness is %v", v.Status.Robustness)
-		return nil
-	}
-
 	// cannot process migrate when upgrading
-	if c.isVolumeUpgrading(v) {
+	if isVolumeUpgrading(v) {
 		log.Warn("Skip the migration processing since the volume is being upgraded")
 		return nil
 	}
 
-	// the only time there should be more then 1 engines is when we are migrating or upgrading
-	// if there are more then 1 and we no longer have a migration id set we can cleanup the extra engine
 	if v.Spec.MigrationNodeID == "" {
+		// The volume attachment controller has stopped the migration (if one was ever started). We must clean up any
+		// extra engines/replicas and leave the volume in a "good" state.
+
+		// The only time there should be more then one engines is when we are migrating or upgrading. If there are more
```

Comment:
NIT: Looks like volume should have 1 engine during upgrading. 2 engines only during migrating

Reply:
Interesting! Looks like you are correct.